frontend/catalyst/jax_primitives.py

```diff
@@ -1650,36 +1650,80 @@ def _hamiltonian_lowering(jax_ctx: mlir.LoweringRuleContext, coeffs: ir.Value, *
     return HamiltonianOp(result_type, coeffs, terms).results
 
 
+#
+# Measurement primitives
+#
+
+
+def shots_is_static(shots: ir.Value):
+    """
+    In many measurement primitives, `shots` might be dynamic or static
+    Hence the corresponding measurement op's result shape might be
+    dynamic or static.
+    It is thus necessary to judge whether a shots value is dynamic or static.
+
+    Static shots will be binded with integer literals, e.g.
+    `sample_p.bind(obs, 5)`
+    These will be stablehlo.constant ops in mlir.
+    """
+    return shots.owner.name == "stablehlo.constant"
+
+
+def get_static_shots_literal_value(shots: ir.Value):
+    """
+    Return the integer literal shots value from a static shots.
+    """
+    assert shots_is_static(shots)
+    return shots.owner.attributes["value"].get_splat_value().value
+
+
 #
 # sample measurement
 #
 @sample_p.def_abstract_eval
 def _sample_abstract_eval(obs, shots, shape):
     assert isinstance(obs, AbstractObs)
 
+    # TODO: restructure these checks now that `shots` is a proper argument instead of an attribute
+    # Is `shape` even necessary now?
     if obs.primitive is compbasis_p:
-        assert shape == (shots, obs.num_qubits)
+        # assert shape == (shots, obs.num_qubits)
+        assert shape[1] == obs.num_qubits
     else:
-        assert shape == (shots,)
+        # assert shape == (shots,)
+        assert len(shape) == 1
 
-    return core.ShapedArray(shape, jax.numpy.float64)
+    return core.DShapedArray(shape, jax.numpy.dtype("float64"))
 
 
 @sample_p.def_impl
 def _sample_def_impl(ctx, obs, shots, shape):  # pragma: no cover
     raise NotImplementedError()
 
 
-def _sample_lowering(jax_ctx: mlir.LoweringRuleContext, obs: ir.Value, shots: int, shape: tuple):
+def _sample_lowering(
+    jax_ctx: mlir.LoweringRuleContext, obs: ir.Value, shots: ir.Value, shape: tuple
+):
+    # Note: result shape of sample op is (shots, number_of_qubits)
     ctx = jax_ctx.module_context.context
     ctx.allow_unregistered_dialects = True
 
-    i64_type = ir.IntegerType.get_signless(64, ctx)
-    shots_attr = ir.IntegerAttr.get(i64_type, shots)
     f64_type = ir.F64Type.get()
-    result_type = ir.RankedTensorType.get(shape, f64_type)
 
-    return SampleOp(result_type, obs, shots_attr).results
+    if shots_is_static(shots):
+        result_shape = (get_static_shots_literal_value(shots), shape[1])
+        assert result_shape == shape
+        i64_type = ir.IntegerType.get_signless(64, ctx)
+        shots = TensorExtractOp(i64_type, shots, []).result
+
+    else:
+        # TODO: use actual dynamic shape tensor type API
+        # Somehow this magic number means dynamically sized dimensions in jax
+        result_shape = (-9223372036854775808, shape[1])
```

Comment:
How do they use this number? Is it defined globally?

Reply:
It is now no longer used, and we use the proper `ir.ShapedType.get_dynamic_size()` instead (which just returns this number).